runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java

```diff
@@ -43,26 +48,46 @@
  * artifacts in memory..
  */
 public class InMemoryArtifactStagerService extends ArtifactStagingServiceImplBase {
-  private final Map<ArtifactMetadata, byte[]> artifactBytes;
+  private final ConcurrentMap<ArtifactMetadata, byte[]> artifactBytes;
   private Manifest manifest;
 
-  public InMemoryArtifactStagerService() {
-    artifactBytes = new HashMap<>();
+  private final AtomicInteger activePuts = new AtomicInteger();
+
+  InMemoryArtifactStagerService() {
+    artifactBytes = new ConcurrentHashMap<>();
   }
 
   @Override
   public StreamObserver<ArtifactApi.PutArtifactRequest> putArtifact(
       StreamObserver<ArtifactApi.PutArtifactResponse> responseObserver) {
-    return new BufferingObserver(responseObserver);
+    BufferingObserver bufferingObserver = new BufferingObserver(responseObserver);
+    activePuts.getAndIncrement();
+    return bufferingObserver;
   }
 
   @Override
   public void commitManifest(
       ArtifactApi.CommitManifestRequest request,
       StreamObserver<ArtifactApi.CommitManifestResponse> responseObserver) {
     this.manifest = request.getManifest();
-    responseObserver.onNext(CommitManifestResponse.getDefaultInstance());
-    responseObserver.onCompleted();
+    while (activePuts.get() != 0) {
+      // Wait for existing calls to complete.
```

Comment:
Can there be new calls after this loop?

Reply:
Yes; however, I generally expect that tests (which is where this should be used) will be sufficiently responsive to 

I can add a AtomicBoolean committing and throw if I a put tries to start after we start committing. Should I?